mux/mux.go

```diff
@@ -19,10 +19,30 @@ type MuxedStream interface {
 	io.Reader
 	io.Writer
 
-	// Close closes the stream for writing. Reading will still work (that
-	// is, the remote side can still write).
+	// Close closes the stream.
+	//
+	// * Any buffered data for writing will be flushed.
+	// * Future reads will fail.
+	// * Any in-progress reads/writes will be interrupted.
+	//
+	// Close may be asynchronous and _does not_ guarantee receipt of the
+	// data.
 	io.Closer
 
+	// CloseWrite closes the stream for writing but leaves it open for
+	// reading.
+	//
+	// CloseWrite does not free the stream, users must still call Close or
+	// Reset.
+	CloseWrite() error
+
+	// CloseRead closes the stream for writing but leaves it open for
+	// reading.
+	//
+	// CloseRead does not free the stream, users must still call Close or
```

Comment:
How about a `CloseRead()` followed by a `CloseWrite()`? Should the contract enforce the release of resources?

Reply:
A nice implementation should but I'm not sure if we want to require that. This is kind of like `shutdown(SHUT_RD)` followed by `shutdown(SHUT_WR)`, the user still needs to call `close()`.